src/builder/value_parser.rs

```diff
@@ -637,6 +637,34 @@ pub trait TypedValueParser: Clone + Send + Sync + 'static {
     ) -> Option<Box<dyn Iterator<Item = crate::PossibleValue<'static>> + '_>> {
         None
     }
+
+    /// Restrict the valid values for the parser to a range.
+    ///
+    /// See [`InRange`].
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use clap::builder::TypedValueParser;
+    ///
+    /// let mut cmd =clap::Command::new("raw")
+    ///     .arg(
+    ///         clap::Arg::new("port")
+    ///             .long("port")
+    ///             .value_parser(clap::value_parser!(u16).in_range(3000..4000))
+    ///             .action(clap::ArgAction::Set)
+    ///             .required(true)
+    ///     );
+    /// let m = cmd.try_get_matches_from_mut(["cmd", "--port", "3001"]).unwrap();
+    /// let port: u16 = *m.get_one("port").expect("required");
+    /// assert_eq!(port, 3001);
+    /// ```
+    fn in_range<R: RangeBounds<Self::Value>>(self, range: R) -> InRange<Self>
```

Comment:
Unsure how I feel about this being on the trait.

Reply:
Yeah, I don't love it either. Some other options would be a standalone function or macro (ex `range_parser!(u16: 3000..4000)`). Or maybe creating a different trait for types that can be restricted to a range?